samtranslator/open_api/open_api.py

```diff
@@ -376,6 +379,75 @@ def add_tags(self, tags):
+        allowed_credentials=None,
+    ):
+        """
+        Add CORS configuration to this Api to _X_APIGW_CORS header in open api definition
+
+        Following this guide:
+        https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-cors.html
+        https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-apigatewayv2-api-cors.html
+
+        :param list/dict allowed_origins: Comma separate list of allowed origins.
+            Value can also be an intrinsic function dict.
+        :param list/dict allowed_headers: Comma separated list of allowed headers.
+            Value can also be an intrinsic function dict.
+        :param list/dict allowed_methods: Comma separated list of allowed methods.
+            Value can also be an intrinsic function dict.
+        :param list/dict expose_headers: Comma separated list of allowed methods.
+            Value can also be an intrinsic function dict.
+        :param integer/dict max_age: Maximum duration to cache the CORS Preflight request. Value is set on
+            Access-Control-Max-Age header. Value can also be an intrinsic function dict.
+        :param bool/None allowed_credentials: Flags whether request is allowed to contain credentials.
+        """
+        ALLOW_ORIGINS = "allowOrigins"
+        ALLOW_HEADERS = "allowHeaders"
+        ALLOW_METHODS = "allowMethods"
+        EXPOSE_HEADERS = "exposeHeaders"
+        MAX_AGE = "maxAge"
+        ALLOW_CREDENTIALS = "allowCredentials"
+        cors_headers = [ALLOW_ORIGINS, ALLOW_HEADERS, ALLOW_METHODS, EXPOSE_HEADERS, MAX_AGE, ALLOW_CREDENTIALS]
+        cors_configuration = self._doc.get(self._X_APIGW_CORS, dict())
+
+        # intrinsics will not work if cors configuration is defined in open api and as a property to the HttpApi
+        if allowed_origins and is_intrinsic(allowed_origins):
+            cors_configuration_string = json.dumps(allowed_origins)
+            for header in cors_headers:
+                keyword = header[0].upper() + header[1:]
+                count = cors_configuration_string.count(keyword)
+                cors_configuration_string = cors_configuration_string.replace(keyword, header, count)
+
+            cors_configuration_dict = json.loads(cors_configuration_string)
+            cors_configuration.update(cors_configuration_dict)
+
+        else:
+            if allowed_origins:
+                cors_configuration[ALLOW_ORIGINS] = allowed_origins
+            if allowed_headers:
+                cors_configuration[ALLOW_HEADERS] = allowed_headers
+            if allowed_methods:
+                cors_configuration[ALLOW_METHODS] = allowed_methods
+            if expose_headers:
+                cors_configuration[EXPOSE_HEADERS] = expose_headers
+            if max_age is not None:
+                cors_configuration[MAX_AGE] = max_age
+            if allowed_credentials is True:
+                cors_configuration[ALLOW_CREDENTIALS] = allowed_credentials
```

Comment:
If I understand this right, we are enforcing the following behaviors for all properties (except `max_age` and `allowed_credentials`):
* Globals: globals are merged, which means lists are added to each other (between global and resource-level `CorsConfiguration` properties, not dealing with the OpenApi definition yet)
* OpenApi merging: when merging with OpenApi, the CORS section is replaced (instead of merged) for any conflicting keys. This means that `CorsConfiguration` overrides any pre-existing cors configurations in the open api doc.

Is this correct ^^?

Reply:
Could we have a test that combines these three parts? Globals + Resource-Level + OpenApi for a Cors property.

Reply:
I have a test case for Local and OpenApi merging here https://github.com/awslabs/serverless-application-model/pull/1381/files#diff-41c41f1a595c5e3a950fee593ef28b21. Yes, you are correct. `CorsConfiguration` overrides any pre-existing config in open api.

I will add a Globals as well to this testcase and update the PR.